source/decorated.d

```diff
@@ -135,4 +143,38 @@ unittest
         assert(bar() == "D roks: Hello Decorated");
         assert(baz() == "D roks: Hello Decorated");
     }
+}
+
+unittest
+{
+    auto cache(F)(F f) @safe pure nothrow
+    {
+        import std.functional : memoize;
+        alias mem = memoize!f;
+        return &mem;
+    }
+
+    // https://issues.dlang.org/show_bug.cgi?id=22694
+    struct S
+    {
+        static size_t expensive; // simulates an expensive algorithm
+
+        @cache
+        mixin decorated!("fib", (size_t n = 1)
+        {
+            expensive++; // simulates an expensive algorithm
+            return expensive;
```

Comment:
Why not actually adding fibonacci? Otherwise rename it and do the following change:

```suggestion
            return expensive++; // simulates an expensive algorithm
```

Reply:
First, Fibonacci is not the best example to use here, and second, this does not need an expensive algorithm running. I'm not testing performance, just testing the practicality of the feature.

Because Fibonacci is a pure algorithm it'll always return the same value given the same input. Someone actually reading the unit tests to understand how it works won't know the difference between both calls. Here I'm enforcing that expensive will increment with every call, however, because it is cached, it won't.

And this is also why I won't make this change, I want to emphasize that the call is skipped. Adding that fancy code will just confuse the reader.

Reply:
Sure, I can agree about the unnecessary complexity, but then please do the latter.

Reply:
done